src/server/project.ts

```diff
@@ -1577,7 +1577,7 @@ namespace ts.server {
             }
         }
 
-        protected enablePlugin(pluginConfigEntry: PluginImport, searchPaths: string[], pluginConfigOverrides: Map<any> | undefined) {
+        protected async enablePlugin(pluginConfigEntry: PluginImport, searchPaths: string[], pluginConfigOverrides: Map<any> | undefined) {

```

Comment:
If `enablePlugin` is async, we may need to propagate `async` to its callers such as `Project.enableGlobalPlugins` and `ConfiguredProject.enablePluginsWithOptions`, and so on. By making this `async` and not handling the response, any errors loading synchronous plugins in NodeJS will be swallowed and reported as unhandled rejections. We might want to have a separate function for async plugin loading so that sync plugin loading isn't affected by this behavior.

I'm also concerned about possible race conditions related to:
- plugin load order
- cache invalidation in the language server

Reply:
Good point about the races. One thing that helps us is that we either have sync or async plugins, not both at the same time. But in the async case, I will need to propagate awaits up to `enableGlobalPlugins` to make sure the plugins are loaded in the correct order. Looking into this now

For the separate functions, do you mean having an `enablePluginSync` and `enablePluginAsync`?  I quickly tried looking into this but was concerned this would push the problem up to the caller. Could you share an example of what you think this would look like

Reply:
Making `enableGlobalPlugins` async pushes the problem with unhandled rejections up to the constructors for `InferredProject` and `ExternalProject`, which can't be made async because they're constructors, as well as `ProjectService.loadConfiguredProject`, which expects to be able to read external files added by plugins on the next line. 

For the `InferredProject` and `ExternalProject` constructors, we might need to have plugin loading occur as a separate step after construction. For other cases we'd likely need to propagate `async` fairly far up the stack.

Reply:
Ok. I think I'll need some help with this because I'm not very familiar with the TS code base. Should we schedule some time to pair on this or would you be able to look into the problem?